toolchain/driver/clang_runtimes.cpp

```diff
@@ -0,0 +1,403 @@
+    it = std::lower_bound(obj_dirs_.begin(), it, obj_dir_path);
+    if (*it != obj_dir_path) {
+      it = obj_dirs_.insert(it, obj_dir_path);
+    }
+    obj_dir_path = obj_dir_path.parent_path();
+  }
+  return Success();
+}
+
+auto ClangRuntimesBuilderBase::ArchiveBuilder::CompileMember(
+    llvm::StringRef src_file) -> ErrorOr<llvm::NewArchiveMember> {
+  // Create any obj subdirectories needed for this file.
+  CARBON_RETURN_IF_ERROR(CreateObjDir(src_file.str()));
+
+  std::filesystem::path obj_path =
+      builder_->runtimes_builder_->path() / std::string_view(src_file);
+  obj_path += ".o";
+  std::filesystem::path src_path = srcs_path_ / std::string_view(src_file);
+  CARBON_VLOG("Building `{0}' from `{1}`...\n", obj_path, src_path);
+
+  llvm::SmallVector<llvm::StringRef> args(cflags_);
+
+  // Add language-specific flags based on file extension.
+  if (src_file.ends_with(".c")) {
+    args.push_back("-std=c11");
+  } else if (src_file.ends_with(".cpp")) {
+    args.push_back("-std=c++20");
+  }
+
+  // Collect the additional required flags and dynamic flags for this builder.
+  args.append({
+      "-c",
+      builder_->target_flag_,
+      "-o",
+      obj_path.native(),
+      src_path.native(),
+  });
+  if (!builder_->clang_->RunWithNoRuntimes(args)) {
+    return Error(
+        llvm::formatv("Failed to compile runtime source file '{0}'", src_file));
+  }
+
+  auto obj_result = llvm::NewArchiveMember::getFile(obj_path.native(),
+                                                    /*Deterministic=*/true);
+  if (!obj_result) {
+    return Error(llvm::formatv("Unable to read `{0}` object file: {1}",
+                               src_file,
+                               llvm::fmt_consume(obj_result.takeError())));
+  }
+
+  // Unlink the object file once we've read it. However, we log and ignore
+  // any errors here as they aren't fatal.
+  auto unlink_result = builder_->runtimes_builder_->dir().Unlink(obj_path);
+  if (!unlink_result.ok()) {
+    CARBON_VLOG("Unable to unlink object file `{0}`: {1}\n", obj_path,
+                unlink_result.error());
+  }
+
+  return std::move(*obj_result);
+}
+
+ClangResourceDirBuilder::ClangResourceDirBuilder(
+    ClangRunner* clang, llvm::ThreadPoolInterface* threads,
+    llvm::Triple target_triple, Runtimes* runtimes)
+    : ClangRuntimesBuilderBase(clang, threads, std::move(target_triple)),
+      crt_begin_result_(Error("Never built CRT begin file!")),
+      crt_end_result_(Error("Never built CRT end file!")) {
+  // Ensure we're on a platform where we _can_ build a working runtime.
+  if (target_triple_.isOSWindows()) {
+    result_ =
+        Error("TODO: Windows runtimes are untested and not yet supported.");
+    return;
+  }
+
+  auto build_dir_or_error = runtimes->Build(Runtimes::ClangResourceDir);
+  if (!build_dir_or_error.ok()) {
+    result_ = std::move(build_dir_or_error).error();
+    return;
+  }
+  auto build_dir = *(std::move(build_dir_or_error));
```

Comment:
I'd probably drop the () around the std::move?
```suggestion
  auto build_dir = *std::move(build_dir_or_error);
```